File: Ya2.0/8lesson/C++/C.cpp
#include <iostream>

#include <string>
#include <sstream>
#include <unordered_map>
#include <unordered_set>

using namespace std;

int main()
{
    ios::sync_with_stdio(NULL), cin.tie(0), cout.tie(0);

    int n;
    cin >> n;
    unordered_map<string, string> tree;
    for (int i = 0; i != n - 1; ++i) {
        string child, ancs;
        cin >> child >> ancs;

        tree[child] = ancs;
    }

    string line;
    getline(cin, line);// прожевываем '\n'
    getline(cin, line);

    while (!line.empty()) {
        istringstream ss(line);

        string name1, name2;
        ss >> name1 >> name2;

        unordered_set<string> path;
        string it = name1;
        while (tree.count(it)) {
            path.insert(it);
            it = tree[it];
        }
        path.insert(it); // добавляем корень

        it = name2;
        while (tree.count(it) && !path.count(it)) {
            it = tree[it];
        }

        cout << it << endl;
        getline(cin, line);
    }

    return 0;
}

File: Ya2.0/8lesson/Go/C/C.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func main() {
	cin := bufio.NewReader(os.Stdin)
	cout := bufio.NewWriter(os.Stdout)
	
	var N int
	fmt.Fscan(cin, &N)	

	tree := make(map[string]string, N)

	for i := 0; i != N - 1; i++ {
		var child, ancs string
		fmt.Fscan(cin, &child, &ancs)

		tree[child] = ancs
	}

	sc := bufio.NewScanner(os.Stdin)

	for sc.Scan() {
		if sc.Text() == "" {
			break
		}

		str := strings.Split(sc.Text(), " ")
		name1 := str[0]
		name2 := str[1]

		path := make(map[string]int)
		it := name1
		for it != "" {
			path[it] = 1
			it = tree[it]
		}

		it = name2
		_, flagF := path[it]

		for !flagF {
			it = tree[it]
			_, flagF = path[it]
		}

		fmt.Fprintln(cout, it)
	}
	
	cout.Flush()
}